### PR TITLE
Improve use of protocol

### PR DESCRIPTION
**Pull request by MathiasReker**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | prevent redirect in some cases.
| Type?         | improvement
| Category?     |  CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | /
| How to test?  | /

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11635)

### CODE REVIEW

File: classes/shop/Shop.php [hunk 1]
```diff
@@ -529,7 +529,7 @@ public function getBaseURL($auto_secure_mode = false, $add_base_uri = true)
         }
 
         $url = array();
-        $url['protocol'] = $auto_secure_mode && Tools::usingSecureMode() ? 'https://' : 'http://';
+        $url['protocol'] = Tools::getShopProtocol();
```
Comment (Quetzacoalt91) [inline classes/shop/Shop.php]: I'm unsure this change is a good idea, as we can have a different domain per protocol. I'd like to keep the match.

Comment (MathiasReker) [reply classes/shop/Shop.php]: Fixed.

**Review by Quetzacoalt91 — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
Hello @MathiasReker,

can you explain how we can test it, please?

Mickaël

**Comment by MathiasReker** [comment]
@mickaelandrieu I don't know where is header is set, sorry.  I just made the PR from a theoretic point of view.

**Comment by mickaelandrieu** [comment]
ping @Quetzacoalt91 any chance to set up a shop in https with this pull request?

Before: headers says HTTP
After: headers says HTTPS

When a product doesn't exist and/or when you configure a redirection for a product.

File: classes/shop/Shop.php [hunk 2]
```diff
@@ -369,7 +369,7 @@ public static function initialize()
                         $redirect_header = ($redirect_type == 1 ? 'Found' : 'Moved Permanently');
                         header('HTTP/1.0 ' . $redirect_code . ' ' . $redirect_header);
                         header('Cache-Control: no-cache');
-                        header('Location: http://' . $url);
+                        header('Location: ' . Tools::getShopProtocol() . $url);
```
Comment (Quetzacoalt91) [inline classes/shop/Shop.php:372]: This one covers a wrong domain.

For instance, if you installed your shop locally with the domain `127.0.0.1`, this code will be called by reaching your website from `http(s)://localhost`.

File: classes/shop/Shop.php [hunk 3]
```diff
@@ -441,7 +441,7 @@ public static function initialize()
                 $redirect_code = ($redirect_type == 1 ? '302' : '301');
                 $redirect_header = ($redirect_type == 1 ? 'Found' : 'Moved Permanently');
                 header('HTTP/1.0 ' . $redirect_code . ' ' . $redirect_header);
-                header('Location: http://' . $url);
+                header('Location: ' . Tools::getShopProtocol() . $url);
```
Comment (Quetzacoalt91) [inline classes/shop/Shop.php:444]: To trigger this code, call a shop with URL rewriting enabled through a URL generated without GET parameters.

From instance: Calling `http://<shop_domain>/index.php?idPayment=1337&fc=module&module=cheque&controller=validation&id_lang=1` should trigger this code to redirect you to `http://<shop_domain>/en/module/cheque/validation?idPayment=1337`

**Comment by Quetzacoalt91** [comment]
If the QA is unable to create its shop with SSL, I can have one something for them.

**Comment by Quetzacoalt91** [comment]
Thank you @MathiasReker

**Pull request merged**